scanpy/metrics/_gearys_c.py

```diff
@@ -0,0 +1,286 @@
+    assert choices_made <= 1
+    if choices_made == 0:
+        return adata.X
+    elif is_layer:
+        return adata.layers[layer]
+    elif use_raw:
+        return adata.raw.X
+    elif is_obsm:
+        return adata.obsm[obsm]
+    elif is_obsp:
+        return adata.obsp[obsp]
+    else:
+        raise RuntimeError("You broke it. But how? Please report this.")
+
+
+###############################################################################
+# Calculation
+###############################################################################
+# Some notes on the implementation:
+# * This could be phrased as tensor multiplication. However that does not get
+#   parallelized, which boosts performance almost linearly with cores.
+# * Due to the umap setting the default threading backend, a parallel numba
+#   function that calls another parallel numba function can get stuck. This
+#   ends up meaning code re-use will be limited until umap 0.4.
+#   See: https://github.com/lmcinnes/umap/issues/306
+# * There can be a fair amount of numerical instability here (big reductions),
+#   so data is cast to float64. Removing these casts/ conversion will cause the
+#   tests to fail.
+
+
+@numba.njit(cache=True, parallel=True)
+def _gearys_c_vec(data, indices, indptr, x):
+    W = data.sum()
+    return _gearys_c_vec_W(data, indices, indptr, x, W)
+
+
+@numba.njit(cache=True, parallel=True)
+def _gearys_c_vec_W(data, indices, indptr, x, W):
+    N = len(indptr) - 1
+    x_bar = x.mean()
+    x = x.astype(np.float_)
+
+    total = 0.0
+    for i in numba.prange(N):
+        s = slice(indptr[i], indptr[i + 1])
+        i_indices = indices[s]
+        i_data = data[s]
+        total += np.sum(i_data * ((x[i] - x[i_indices]) ** 2))
+
+    numer = (N - 1) * total
+    denom = 2 * W * ((x - x_bar) ** 2).sum()
+    C = numer / denom
+
+    return C
+
+
+@numba.njit(cache=True, parallel=True)
+def _gearys_c_mtx_csr(
+    g_data, g_indices, g_indptr, x_data, x_indices, x_indptr, x_shape
+):
+    M, N = x_shape
+    W = g_data.sum()
+    out = np.zeros(M, dtype=np.float_)
+    for k in numba.prange(M):
+        x_k = np.zeros(N, dtype=np.float_)
+        sk = slice(x_indptr[k], x_indptr[k + 1])
+        x_k_data = x_data[sk]
+        x_k[x_indices[sk]] = x_k_data
+        x_k_bar = np.sum(x_k_data) / N
+        total = 0.0
+        for i in numba.prange(N):
+            s = slice(g_indptr[i], g_indptr[i + 1])
+            i_indices = g_indices[s]
+            i_data = g_data[s]
+            total += np.sum(i_data * ((x_k[i] - x_k[i_indices]) ** 2))
+        numer = (N - 1) * total
+        # Expanded from 2 * W * ((x_k - x_k_bar) ** 2).sum(), but uses sparsity
+        # to skip some calculations
+        denom = (
+            2
+            * W
+            * (
+                np.sum(x_k_data ** 2)
+                - np.sum(x_k_data * x_k_bar * 2)
+                + (x_k_bar ** 2) * N
+            )
+        )
```

Comment:
This was reformatted ugly, please do

```py
denom_sum = (
    ...
)
C = numer / (2 * W * denom_sum)
```

Reply:
I'd rather use `# fmt: off` above the statement than have a temporary variable. I also don't mind the current one. If you could have it without the temporary variable, what would you want it to look like? Maybe this?

```python
# fmt: off
denom = (
    2 * W
    * (
        np.sum(x_k_data ** 2)
        - np.sum(x_k_data * x_k_bar * 2)
        + (x_k_bar ** 2) * N
    )
)
```

Reply:
This might have broken black.

Reply:
This definitely broke black. Mind if we leave it with the long formatting until that gets resolved?

Reply:
OK!